File: k-diff/kdiff.py
from functools import reduce

def upper_bound(arr, n, value):
    """Find the first index where arr[index] > value (Upper Bound)."""
    low, high = 0, n
    while low < high:
        mid = (low + high) // 2
        high = mid if arr[mid] > value else low + 1
    return low

def count_pairs(arr, n, mid):
    """Count number of pairs with absolute difference <= mid using lambda & reduce."""
    return reduce(lambda acc, i: acc + (upper_bound(arr, n, arr[i] + mid) - (i + 1)), range(n), 0)

def kth_diff(arr, k):
    arr.sort()
    
    # Finding the minimum absolute difference using reduce
    low = reduce(lambda acc, i: min(acc, arr[i+1] - arr[i]), range(len(arr)-1), float('inf'))
    high = arr[-1] - arr[0]  # Maximum absolute difference

    # Binary search to find the k-th smallest absolute difference
    while low < high:
        mid = (low + high) // 2
        if count_pairs(arr, len(arr), mid) < k:
            low = mid + 1
        else:
            high = mid

    return low

# Example usage
arr = [1, 2, 3, 4]
k = 3
print(kth_diff(arr, k))  # Output: 2

File: k-diff/kdiff.java
import java.util.Arrays;
import java.util.Scanner;
import java.util.function.BiFunction;

class kdiff {
    public static void main(String args[]) {
        Scanner s = new Scanner(System.in);
        int k = 3;
        int a[] = {1, 2, 3, 4};
        int n = a.length;
        System.out.println(kthDiff(a, n, k));
    }

    // Returns k-th absolute difference
    static int kthDiff(int a[], int n, int k) {
        // Sort array
        Arrays.sort(a);

        // Minimum and Maximum absolute difference
        int low = Arrays.stream(a).reduce((x, y) -> Math.min(y - x, Integer.MAX_VALUE)).getAsInt();
        int high = a[n - 1] - a[0];

        // Upper bound lambda using binary search
        BiFunction<int[], Integer, Integer> upperbound = (arr, value) -> {
            int pos = Arrays.binarySearch(arr, value);
            return (pos < 0) ? -pos - 1 : pos + 1;
        };

        // Lambda for counting pairs
        BiFunction<int[], Integer, Integer> countPairs = (arr, mid) -> {
            int res = 0;
            for (int i = 0; i < arr.length; i++) {
                int ub = upperbound.apply(arr, arr[i] + mid);
                res += (ub - (i + 1));
            }
            return res;
        };

        // Binary search for k-th absolute difference
        while (low < high) {
            int mid = (low + high) >> 1;
            if (countPairs.apply(a, mid) < k)
                low = mid + 1;
            else
                high = mid;
        }

        return low;
    }
}
